### ISSUE TITLE
Support read and write directly using files.

### ISSUE THREAD

**Issue by isoos** [brendan-duncan/archive#35]
Large archive files take a lot of memory when the buffer needs to be built in memory. Please add support to read and write using files, without the need to build up the buffer in memory.

**Comment by brendan-duncan** [comment]
I'll add file streaming support, with a controllable memory cache. Because tar files are compressed as a whole (tar.gz), rather than per file within the archive like zip, to avoid keeping everything in memory, I'll probably need to use an intermediate temp file to store the uncompressed tar. Does that sound reasonable?

**Comment by brendan-duncan** [comment]
Looking at this further, I quickly realized io.File does not have a method to read N bytes, and the only way to do this is to finally stop resisting and learn those Stream and Converter classes. From what I can tell with the examples I've seen of the Stream class, this will not be a straight forward change.  This task will take longer than I was anticipating, I'm not sure how much longer.

**Comment by isoos** [comment]
You can have more control over that with: https://api.dartlang.org/stable/1.23.0/dart-io/RandomAccessFile/read.html

**Comment by brendan-duncan** [comment]
Excellent, thanks for the pointer!

**Comment by brendan-duncan** [comment]
I have disk->disk tar and gzip encoding working, adjustable size for the input read cache.
With this it's possible to create a tar.gz of a directory with little to no memory overhead.

Still working on decoding streaming, and trying to clean up the interface. It's a challenge keeping it backwards compatible and not introducing dart:io dependency to the main package, but seems to be ok so far.

**Comment by thosakwe** [comment]
Are you accepting PR's?

**Comment by isoos** [comment]
Hi @brendan-duncan, do you have any ETA for this?

**Comment by brendan-duncan** [comment]
Sorry, I don't. A new job is blocking my ability to put any work into this.
I'm not sure if that will change. Maybe someone else will be willing to
take it over.

**Comment by OsamaAbbas** [comment]
> I have disk->disk tar and gzip encoding working

@brendan-duncan Could you please explain how to achieve this part?

**Comment by iulian0512** [comment]
i have the same issue, i tried in many ways but i cannot get this library to compress a 1gb file without allocating on the heap 1gb ... which leads on many devices to out of memory and a fatal app crash.

**Comment by brendan-duncan** [comment]
You're right. I spent time optimizing memory usage for decoding, seems I neglected encoding. I'll need to refactor that code a bit to use streams. Not entirely sure when I'll be able to get to it.

**Comment by iulian0512** [comment]
> You're right. I spent time optimizing memory usage for decoding, seems I neglected encoding. I'll need to refactor that code a bit to use streams. Not entirely sure when I'll be able to get to it.

it's a good library, I appreciate your efforts.

**Comment by brendan-duncan** [comment]
I started playing around with it a bit while I had a few minutes. It should at least not read the entire input file into memory now, but it still needs to store the compressed data in memory until I figure out how to get around that.
